Xbrl metadata restructuring

zaneselvans opened the pull request:
## PR Overview

Use the new per-table XBRL metadata structure, rather than having to look up every column individually.

Now the XBRL instant/duration metadata is extracted based on the `Ferc1Settings` object, and passed in to the individual `TableTransformer` class when it's instantiated. The JSON structure is processed into a dataframe that combines the instant and duration table metadata, so it can be applied to the final concatenated XBRL+DBF table. 

Got rid of some unnecessary metadata column renaming that was happening in the metadata structures, which simplified things substantially.  The only table that has a special bespoke metadata transform now is `plant_in_service_ferc1`, and this has also been simplified, as @cmgosnell noted that the left merge which happens in `merge_xbrl_metadata()` already effectively removes the extraneous row categories.

Also renamed `merge_metadata_xbrl` to `merge_xbrl_metadata` since it's actually metadata for all of the data, and we have a naming convention where `_dbf`, `_xbrl`, `_instant_xbrl` and `_duration_xbrl` suffixes on the methods indicate what portion of the transformation it's meant to be used in.

## Questions / Comments:
* `ElectricEnergySourcesFerc1TableTransformer.drop_invalid_rows()` was dropping **all** rows because its parameters had not been updated to use `energy_mwh` rather than `energy_source_mwh`. We need validations that will catch this kind of thing and fail loudly if we mess up -- even if they're only going to fail in the nightly builds (which are now happening and passing... nightly!)
* Why do we need to drop the disposition columns in the sources `process_dbf()` but we **don't** need to drop the sources columns in the dispositions `process_dbf()`? Shouldn't these transformers be symmetric?
* There was no `record_id` field in the electric energy sources / dispositions tables so I added one... was there a reason it got left out before?
* What is the thinking behind renaming `xbrl_factoid` within `merge_xbrl_metadata()`? Does the rename need to be separately parameterized? Based on current examples it seems like we can always rename the `xbrl_factoid` column to whatever column is named in `MergeXbrlMetadata.on`.

Closes #2110

## PR Checklist
* [x] Merge in the most recent version of `dev`.
* [ ] Verify that all of the CI checks on your PR are passing.
* [ ] Good docstrings for any new modules, classes, functions, or methods
* [ ] Unit tests for new functions or classes.
* [ ] Update the release notes, including references to relevant issues & PRs.
* [ ] Do your own review of the PR. Add comments highlighting areas that need attention.

zaneselvans commented on src/pudl/metadata/resources/ferc1.py:
```diff
@@ -32,6 +32,7 @@
                 "energy_source_type",
                 "row_type_xbrl",
                 "energy_mwh",
+                "record_id",
```
Did these tables not have `record_id` columns for some reason?

cmgosnell replied:
whooops!

zaneselvans replied:
🙈

zaneselvans commented on src/pudl/transform/params/ferc1.py:
```diff
@@ -2992,7 +2992,7 @@
             {
                 "invalid_values": [pd.NA, np.nan, ""],
-                "required_valid_cols": ["energy_source_mwh"],
+                "required_valid_cols": ["energy_mwh"],
```
This naming inconsistency resulted in all of the data being dropped, and we were not catching it. We need to be adding some sanity checks to the data validations so at least the nightly builds catch this kind of thing.

cmgosnell replied:
i noticed this in one of my recent table transforms. think `drop_invalid_rows` should also just let you know if a `required_valid_cols` isn't in the table at all

zaneselvans replied:
There was a warning in the logging output, but it didn't fail.

zaneselvans replied:
More generally I think we should be adding data validation expectations to these tables as we add them to the database -- the null columns check would have found this, as would a validation indicating the expected number of rows.

cmgosnell replied:
i 100% agree with you here and also we have so many tables to integrate in the next few weeks. I would like to be able to add data validations in a thoughtful way and I don't think we actually have time for that right now.

zaneselvans commented on tox.ini:
```diff
@@ -106,7 +106,7 @@ description = Run all the software unit tests.
     test
 commands =
-    pytest {posargs} {[testenv]covargs} \
+    pytest {tty:--color=yes} {posargs} {[testenv]covargs} \
```
Tox 4.0 got rid of our nice color pytest output, and this brings it back.

zaneselvans commented on src/pudl/extract/ferc1.py:
```diff
@@ -758,10 +758,10 @@ def extract_xbrl(
     for pudl_table in ferc1_settings.tables:
-        # TODO: Raise exception once XBRL tables are fully integrated
-        # For now skip because map is not defined for all pudl tables
+        if pudl_table not in TABLE_NAME_MAP:
+            raise ValueError(f"{pudl_table} not found in the list of known tables.")
```
I went ahead and added these checks for real since we have all the basic pudl tables in now.

zaneselvans commented on src/pudl/transform/ferc1.py:
```diff
@@ -2670,7 +2638,7 @@ class TransmissionFerc1TableTransformer(Ferc1AbstractTableTransformer):
 
 
-class ElectricEnergyAccountSourcesFerc1TableTransformer(Ferc1AbstractTableTransformer):
+class ElectricEnergySourcesFerc1TableTransformer(Ferc1AbstractTableTransformer):
```
Just standardized this to match the table name and the disposition table transformer.

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -519,28 +520,25 @@ class Ferc1AbstractTableTransformer(AbstractTableTransformer):
-    xbrl_metadata_normalized: pd.DataFrame = pd.DataFrame()
-    """A semi-normalized dataframe containing table-specific XBRL metadata."""
+    xbrl_metadata: pd.DataFrame
+    """Dataframe combining XBRL metadata for both instant and duration table columns."""
```
Dataframe **of** XBRL metadata for both instant and duration table columns?

cmgosnell replied:
why does this need to be here as well as defined in `__init__`?

zaneselvans replied:
I think it's generally better to define what attributes a class has up front than to randomly stash stuff in a class attribute you make up as you go along.  This is a statement about the kinds of things one should expect to find inside of an object of type `Ferc1AbstractTableTransformer` (which allows it to be documented as part of the class) and also allows tab-completion based on the declared type.

I should set a default value of an empty dataframe too.

cmgosnell replied:
okay!!

cmgosnell commented on src/pudl/extract/ferc1.py:
```diff
@@ -758,10 +758,10 @@ def extract_xbrl(
+            raise ValueError(f"{pudl_table} not found in the list of known tables.")
         if "xbrl" not in TABLE_NAME_MAP[pudl_table]:
-            continue
+            raise ValueError(f"No XBRL tables have been associated with {pudl_table}.")
```
I could actually imagine cases in which we don't have an XBRL table or a DBF table input for all pudl tables. But i think this is fine for now bc this is our current pattern

cmgosnell reviewed (approved):
looks good! obviously go track down and fix the tests that depend on the old metadata structure. but the existing changes look great. I'm glad you were indeed able to delete the pis metadata pattern based drop columns and only use one version of the xbrl factiod name

zaneselvans commented:
* The `ElectricEnergySources` disposition column dropping cruft was not doing anything -- I got exactly the same dataframe out with and without it, so I removed it.
* Added basic sanity checks to the FERC1 table transformer `transform_end()` method -- it looks for empty dataframes or any 100% null columns.
* `drop_invalid_rows()` will raise an exception if any required valid columns are not present in the dataframe.
* Created a list of common invalid plant names which is now shared across all of the plant table transforms.
